### DISCUSSION TITLE
Query MySQL full-text index set on multiple columns

### DISCUSSION THREAD

**Discussion by Zmaon** [mikro-orm/mikro-orm#4558] [General]
Hello,

I'm trying to query a full-text search index set on multiple columns in MySQL. I can set it in the entity like this:

```typescript
@Index({
  properties: ['first_name', 'last_name', 'email'],
  type: 'fulltext',
})
@Entity({ tableName: 'users' })
export default class User {
  @Property()
  firstName!: string;

...
}
```

With MySQL, if I created my index on 3 columns it's required to query those 3 columns at once in the full-text search query.
I can't figure out how to do it without a QueryBuilder, I guess it would require something like that:

```typescript
const results = await this.userRepository.find({
  ['first_name', 'last_name', 'email']: { $fulltext: 'john doe' },
});
```

Is there a solution?

**Comment by B4nan** [comment]
What query do you want to produce? There is one trick that might work for you, you can use `expr` helper with an array which should produce a tuple search.

```ts
const results = await this.userRepository.find({
  [expr(['firstName', 'lastName', 'email'])]: { $fulltext: 'john doe' },
});
```

**Comment by Zmaon** [comment]
The query I would like to produce is:
```SQL
Select * from users where MATCH(first_name, last_name, email) AGAINST ('john doe' IN BOOLEAN MODE);
```

I've got this error when trying to use `expr()` with a `$fulltext `:

`"message": "Cannot use $fulltext operator on e0.first_name~~~last_name~~~email, property not found"`

**Reply by B4nan** [reply]
Hmm I see, this would need some adjustments to the fulltext operator handling. But it could be rather simple, I see the `prop` validation is there only because of the column type for postgres.

But I am not entirely sure how all of this should be handled in various drivers, it would need a bit of research.

cc @jsprw as he was contributing this feature

**Reply by Zmaon** [reply]
A colleague is having a look to see if it's trivial to add support for this in Mikro-orm. For the `where` part it seems doable, but  we would also need to be able to sort by the same expression in order to have results sorted by relevance. In SQL it would be like this:
```SQL
Select * from users 
    where MATCH(first_name, last_name, email) AGAINST ('john doe' IN BOOLEAN MODE)
    order by MATCH(first_name, last_name, email) AGAINST ('john doe' IN BOOLEAN MODE) DESC;
```

Do you see a way to specify the `QueryOrderMap` so that it does that?

**Reply by B4nan** [reply]
As a workaround, you could try to define a formula like that, it should be possible to query and order by it and the expression will be inlined to both clauses.
